stargazer.js

```diff
@@ -0,0 +1,67 @@
+import Alt from './'
+import { createStore, datasource } from './utils/decorators'
+import axios from 'axios'
+
+const alt = new Alt()
+
+const StargazerActions = alt.generateActions(
+  'fetchingUsers',
+  'usersReceived',
+  'failed'
+)
+
+const StargazerSource = {
+  fetchUsers: {
+    fetch(state) {
+      const url = `https://api.github.com/repos/${state.user}/${state.repo}/stargazers`
+      return axios({ url }).then(response => response.data)
+    },
+
+    cache(state) {
+      return state.users.length ? state.users : null
+    },
+
+    loading: StargazerActions.fetchingUsers,
+
+    success: StargazerActions.usersReceived,
+
+    error: StargazerActions.failed
+  }
+}
+
+@createStore(alt)
+@datasource(StargazerSource)
+class StargazerStore {
+  constructor() {
+    this.user = 'goatslacker'
+    this.repo = 'alt'
+    this.users = []
+    this.errorMessage = null
+    this.isLoading = false
+
+    this.bindListeners({
+      loading: StargazerActions.fetchingUsers,
+      receivedUsers: StargazerActions.usersReceived,
+      failed: StargazerActions.failed
+    })
+  }
+
+  loading() {
+    console.log('STORE: starting to load users')
+    this.isLoading = true
+  }
+
+  failed(e) {
+    console.log('STORE: Uh oh')
+    this.errorMessage = e.statusText || String(e)
+  }
+
+  receivedUsers(users) {
+    console.log('STORE: got users')
+    this.users = users
+    this.errorMessage = null
+  }
+}
+
+StargazerStore.listen((state) => console.log('CHANGED', state.users.length))
+StargazerStore.fetchUsers()
```

Comment:
I wouldn't mind adding an HttpProvider as an add-on which does automatic batching of requests, say you call a request for URL: 'foo/bar' and then another component tries to fetch from the same URL we can just fire the one remote request and respond to both call-sites when that single request comes back.

I'd also love to provide a caching add-on to automate caching of things.

Aborting requests could also be part of HttpProvider. I think we need to take control of the http request implementation in order to be able to provide these features.

Right now this API makes it flexible to be able to build things on top of it. I'd love to hear thoughts on if its flexible enough.